056-kafka-roller-kraft.md

```diff
@@ -0,0 +1,158 @@
+against a Strimzi cluster in KRaft mode.
+The proposal describes the checks the KafkaRoller should take, how it should perform 
+those checks and in what order but does not discuss exactly how the KafkaRoller works.
+This proposal is expected to apply to both the current KafkaRoller and a future iteration of the KafkaRoller.
+
+This proposal assumes that liveness/readiness of nodes is as described in [proposal #46](https://github.com/strimzi/proposals/blob/main/046-kraft-liveness-readiness.md) 
+and [PR #8892](https://github.com/strimzi/strimzi-kafka-operator/pull/8892).
+
+
+## Current situation
+
+When operating on a ZooKeeper based cluster the KafkaRoller has the behaviour 
+described below.
+
+### Order to roll nodes
+Roll in order:
+ 1. Unready brokers
+ 2. Ready brokers
+ 3. Current controller
+
+### Triggers
+The following are some of the triggers that roll a broker:
+- Pod or its StrimziPodSet is annotated for manual rolling update
+- Pod is unready
+- Broker's configuration has changed
+
+### Rollability checks
+ 1. Does not force roll a broker performing log recovery
+ 2. Attempts to connect an admin client to a broker and if it can't connect rolls the broker
+ 3. Force rolls a broker if the pod is in stuck state
+ 4. Rolls a broker if the configuration has changed and cannot be updated dynamically
+ 5. Does not roll a broker if doing so would take the in-sync replicas count below `min.insync.replicas`
+
+
+#### Unready pod
+If pod is unready, KafkaRoller awaits its readiness before doing anything. If it hasn't become ready still, it checks whether it is stuck and considers to force roll. A pod is considered stuck if it is in one of following states:
+- `CrashLoopBackOff`
+- `ImagePullBackOff`
+- `ContainerCreating`
+- `Pending` and `Unschedulable`
+
+If pod is not stuck but unready, the next rollability checks are performed.
+
+
+#### Configuration changes
+ - Retrieves the current Kafka configurations of the broker via admin client and compares it with the desired configurations specified in the Kafka CR. 
+ - Performs dynamic configuration updates if possible
+
+In KRaft mode the KafkaRoller currently skips controller only nodes, but performs the above steps on any combined or broker only nodes.
+This is causing a problem in combined mode because if the quorum has not formed due to some of the nodes not being ready 
+the KafkaRoller will still try to contact the broker via the admin client.
+This call fails because the quorum is not formed, so in some cases this results in the cluster being stuck with some nodes 
+in a pending state.
+
+## Motivation
+
+When running in the KRaft mode the controller pods need to be rolled if configuration changes occur, also at the moment the 
+existing logic is blocking the full implementation of liveness and readiness checks as described in [proposal #46](https://github.com/strimzi/proposals/blob/main/046-kraft-liveness-readiness.md).
```

Comment:
would be worth adding that it's also blocking ZooKeeper to KRaft migration where we need controllers to be rolled more times during the process? FYI this is the proposal https://github.com/strimzi/proposals/pull/93